state/protocol/inmem/epoch.go

```diff
@@ -228,33 +250,103 @@ func (es *committedEpoch) DKG() (protocol.DKG, error) {
 	return dkg, err
 }
 
+// startedEpoch is an epoch which has started, but not ended (ie. the current epoch.)
+// It has all the information of a committedEpoch, plus the epoch's first block height.
+type startedEpoch struct {
+	committedEpoch
+	firstHeight uint64
+}
+
+func (e *startedEpoch) FirstHeight() (uint64, error) {
+	return e.firstHeight, nil
+}
+
+// endedEpoch is an epoch which has ended (ie. the previous epoch). It has all the
+// information of a startedEpoch, plus the epoch's final block height.
+type endedEpoch struct {
+	startedEpoch
+	finalHeight uint64
+}
+
+func (e *endedEpoch) FinalHeight() (uint64, error) {
+	return e.finalHeight, nil
+}
+
 // NewSetupEpoch returns a memory-backed epoch implementation based on an
 // EpochSetup event. Epoch information available after the setup phase will
 // not be accessible in the resulting epoch instance.
-// Error returns:
-// * protocol.ErrNoPreviousEpoch - if the epoch represents a previous epoch which does not exist.
-// * protocol.ErrNextEpochNotSetup - if the epoch represents a next epoch which has not been set up.
-// * state.ErrUnknownSnapshotReference - if the epoch is queried from an unresolvable snapshot.
+// No errors are expected during normal operations.
 func NewSetupEpoch(setupEvent *flow.EpochSetup) (*Epoch, error) {
 	convertible := &setupEpoch{
 		setupEvent: setupEvent,
 	}
-	return FromEpoch(convertible)
+	epoch, err := FromEpoch(convertible)
+	// since we are passing in a concrete service event, no errors are expected
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error constructing setup epoch from service event: %s", err.Error())
+	}
+	return epoch, nil
 }
 
 // NewCommittedEpoch returns a memory-backed epoch implementation based on an
-// EpochSetup and EpochCommit event.
-// Error returns:
-// * protocol.ErrNoPreviousEpoch - if the epoch represents a previous epoch which does not exist.
-// * protocol.ErrNextEpochNotSetup - if the epoch represents a next epoch which has not been set up.
-// * protocol.ErrEpochNotCommitted - if the epoch has not been committed.
-// * state.ErrUnknownSnapshotReference - if the epoch is queried from an unresolvable snapshot.
+// EpochSetup and EpochCommit events.
+// No errors are expected during normal operations.
 func NewCommittedEpoch(setupEvent *flow.EpochSetup, commitEvent *flow.EpochCommit) (*Epoch, error) {
 	convertible := &committedEpoch{
 		setupEpoch: setupEpoch{
 			setupEvent: setupEvent,
 		},
 		commitEvent: commitEvent,
 	}
-	return FromEpoch(convertible)
+	epoch, err := FromEpoch(convertible)
+	// since we are passing in a concrete service event, no errors are expected
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error constructing committed epoch from service events: %s", err.Error())
```

Comment:
Why do we format error as string? Usually we pass error itself.

Reply:
same question. Please see [my suggestion below](https://github.com/onflow/flow-go/pull/3903#discussion_r1109243816)

Reply:
See related discussion here: https://github.com/onflow/flow-go/pull/3903#discussion_r1113512979